File: vreader/Services/Backup/BlobPath.swift
// Purpose: Maps (format, sha256, byteCount) ↔ WebDAV-safe blob path for
// feature #46's content-addressed library blob store.
//
// Path layout: VReader/books/<format>/<sha256>_<byteCount>.<canonical-ext>
//
// Why not the raw fingerprintKey ("epub:abc:1024"): colons are not safe
// across all WebDAV servers — Apache/mod_dav passes them through but some
// proxies, S3-style adapters, and Windows-backed shares mangle them. The
// blob path uses only [0-9a-f_./], leaves the canonical fingerprintKey
// inside the manifest where it's pure data.
//
// Per-format subdirectory ("books/<format>/") keeps directory listings
// bounded as the library grows — most WebDAV servers throttle enumeration
// over ~1000 entries, and sharding by 5 known formats buys headroom.
//
// originalExtension is intentionally NOT in the blob path — different source
// extensions (.mobi, .prc, .azw) for the same SHA-256 content collapse to
// the same blob (canonical .azw3 extension). The original extension travels
// in the library manifest, not the path.
//
// @coordinates-with: BackupBookProjection (PersistenceActor+Backup.swift),
//   BackupLibraryManifestEnvelope (BackupSectionDTOs.swift, future WI-2),
//   BookFileMaterializer (future WI-5),
//   dev-docs/plans/20260503-feature-46-materializing-restore.md

import Foundation

enum BlobPath {

    /// Top-level directory for content-addressed book blobs on the WebDAV
    /// server. Pinned by `BlobPathTests.booksRootIsStable` — changing it
    /// invalidates every previously-uploaded backup.
    static let booksRoot = "VReader/books"

    /// Builds the blob path for a (format, sha256, byteCount) triple.
    ///
    /// Returns: `"VReader/books/<format>/<sha256>_<byteCount>.<canonical-ext>"`
    static func make(format: BookFormat, sha256: String, byteCount: Int64) -> String {
        let formatStr = format.rawValue
        let canonicalExt = format.fileExtensions.first ?? formatStr
        return "\(booksRoot)/\(formatStr)/\(sha256)_\(byteCount).\(canonicalExt)"
    }

    /// Parses a blob path back into its components.
    ///
    /// Returns nil if the path doesn't match the expected layout, the format
    /// segment is unknown, the SHA-256 isn't 64 hex chars, or the byte count
    /// isn't a non-negative integer.
    static func parse(_ path: String) -> (format: BookFormat, sha256: String, byteCount: Int64)? {
        // Expected: "VReader/books/<format>/<sha256>_<byteCount>.<ext>"
        let prefix = booksRoot + "/"
        guard path.hasPrefix(prefix) else { return nil }
        let trail = String(path.dropFirst(prefix.count))

        let segments = trail.split(separator: "/", omittingEmptySubsequences: false)
        guard segments.count == 2 else { return nil }
        let formatSegment = String(segments[0])
        let filename = String(segments[1])

        guard let format = BookFormat(rawValue: formatSegment) else { return nil }

        // Strip extension; rest must be "<sha256>_<byteCount>".
        let nsFilename = filename as NSString
        let stem = nsFilename.deletingPathExtension
        guard !stem.isEmpty, stem != filename else { return nil }  // require an extension

        let parts = stem.split(separator: "_", omittingEmptySubsequences: false)
        guard parts.count == 2 else { return nil }
        let sha = String(parts[0])
        let byteStr = String(parts[1])

        guard isValidSHA256Hex(sha), let bytes = Int64(byteStr), bytes >= 0 else { return nil }
        return (format: format, sha256: sha, byteCount: bytes)
    }

    // MARK: - Helpers

    private static func isValidSHA256Hex(_ s: String) -> Bool {
        guard s.count == 64 else { return false }
        return s.allSatisfy { $0.isHexDigit }
    }
}

File: android/app/src/main/kotlin/com/vreader/app/backup/archive/BlobPath.kt
// Purpose: feature #116 WI-2 (#110 Phase 3) — maps (format, sha256, byteCount) ↔ the WebDAV-safe
// content-addressed blob path, byte-for-byte the iOS `BlobPath` (vreader/Services/Backup/
// BlobPath.swift) so an Android backup's blobs land where iOS restore looks and vice-versa.
//
// Layout: VReader/books/<format>/<sha256>_<byteCount>.<canonicalExt>
// The path uses only [0-9a-f_./] (no colons — some WebDAV proxies/shares mangle them); the
// canonical fingerprintKey stays inside the manifest as pure data. originalExtension is NOT in
// the path — .mobi/.prc/.azw for one SHA collapse to the canonical .azw3 blob; the original
// extension travels in the manifest. The BookFormat enum names ARE the canonical extensions
// (epub/pdf/txt/md/azw3 = iOS `BookFormat.fileExtensions.first`).
package com.vreader.app.backup.archive

import vreader.contracts.BookFormat

object BlobPath {
    /** Top-level dir for content-addressed book blobs — pinned (changing it invalidates every
     *  previously-uploaded backup). Matches iOS `BlobPath.booksRoot`. */
    const val BOOKS_ROOT = "VReader/books"

    /** `VReader/books/<format>/<sha256>_<byteCount>.<canonicalExt>`. */
    fun make(format: BookFormat, sha256: String, byteCount: Long): String =
        "$BOOKS_ROOT/${format.name}/${sha256}_$byteCount.${format.name}"

    /** Inverse of [make]; null if the path doesn't match the layout, the format is unknown, the
     *  SHA isn't 64 hex chars, or the byte count isn't a non-negative integer. */
    fun parse(path: String): Triple<BookFormat, String, Long>? {
        val prefix = "$BOOKS_ROOT/"
        if (!path.startsWith(prefix)) return null
        val trail = path.removePrefix(prefix)
        val segments = trail.split("/")
        if (segments.size != 2) return null
        val format = runCatching { BookFormat.valueOf(segments[0]) }.getOrNull() ?: return null
        val filename = segments[1]
        val dot = filename.lastIndexOf('.')
        if (dot <= 0) return null  // require an extension and a non-empty stem
        // The extension MUST be the format's canonical ext — `make` always emits it, so a path
        // like `books/epub/<sha>_10.pdf` or `books/azw3/<sha>_10.mobi` is NOT our layout.
        if (filename.substring(dot + 1) != format.name) return null
        val stem = filename.substring(0, dot)
        val parts = stem.split("_")
        if (parts.size != 2) return null
        val sha = parts[0]
        val bytes = parts[1].toLongOrNull() ?: return null
        if (!isValidSha256(sha) || bytes < 0) return null
        return Triple(format, sha, bytes)
    }

    private fun isValidSha256(s: String): Boolean =
        s.length == 64 && s.all { it in '0'..'9' || it in 'a'..'f' }
}
